src/Orleans/ActivationService/ActivationServiceUtilities.cs

```diff
@@ -0,0 +1,53 @@
+﻿using Orleans.Runtime;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Orleans
+{
+    public static class ActivationServiceUtilities
+    {
+        private const string ActivationServiceNamesItemName = "ActivationServiceNamesItemName";
+        private const string ConstructorParametersItemName = "ConstructorParameters";
+
+        public static ParameterInfo BindToConstructorParameter(IGrainActivationContext grainActivationContext, Type activationServiceType)
+        {
+            List<string> configuredActivationServices;
+            object item;
+            if (grainActivationContext.Items.TryGetValue(ActivationServiceNamesItemName, out item))
+            {
+                configuredActivationServices = (List<string>)item;
+            }
+            else
+            {
+                configuredActivationServices = new List<string>(1);
+                grainActivationContext.Items[ActivationServiceNamesItemName] = configuredActivationServices;
+            }
+            ParameterInfo[] constructorParameters = GetConstructorParameters(grainActivationContext);
+            ParameterInfo parameter = constructorParameters
+                    .Where(pi => activationServiceType == pi.ParameterType) // skip already configured ActivationServices
+                    .FirstOrDefault(pi => !configuredActivationServices.Contains(pi.Name));
+
+            if (parameter == null) throw new InvalidOperationException("Could not find appropriate parameter in constructor");
+
+            configuredActivationServices.Add(parameter.Name);
+            return parameter;
+        }
+
+        private static ParameterInfo[] GetConstructorParameters(IGrainActivationContext grainActivationContext)
+        {
+            object item;
+            if (grainActivationContext.Items.TryGetValue(ConstructorParametersItemName, out item))
+            {
+                return (ParameterInfo[])item;
+            }
+
+            var constructor = grainActivationContext.GrainType.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            if (constructor == null) throw new InvalidOperationException("Could not find appropriate constructor");
+            ParameterInfo[] constructorParameters = constructor.GetParameters();
+            grainActivationContext.Items[ConstructorParametersItemName] = constructorParameters;
```

Comment:
putting it in grain type data would require changes and more allocations to it, when it will probably not be used that frequently. The reality is that this is just caching extra info, so I would just create a static CachedReadConcurrentDictionary<Type,. ParameterInfo[]> in here... it's all static information based on the the type anyway, so there is no risk for of this static dictionary

---
In reply to: [126529497](https://github.com/dotnet/orleans/pull/3208#discussion_r126529497) [](ancestors = 126529497)